### PR TITLE
[storage][pruner] Separete ledger and state pruners

### PR DESCRIPTION
**Pull request by zcchahaha**
### Description
- Refactor pruner/mod.rs into `LedgerPrunerManager` and `StatePrunerManager`.
- In AptosDB, have one instance of `LedgerPrunerManager` and one instance of `StatePrunerManager` instead of having one instance of `Pruner` which managed both ledger pruner and state pruner.
- Threading model is not changed in this PR.
- Next step includes:
  - Refactor the pruners so that when the pruner is not enabled, we don't spawn a thread in the corresponding worker but always initialize PrunerManager to reduce the `Some(...` in code.
  - Change the threading model to have background threads periodically prune ledger and state,
### Test Plan
Existing UTs.

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2360)

### CODE REVIEW

File: storage/aptosdb/src/lib.rs [hunk 1]
```diff
@@ -265,7 +264,8 @@ pub struct AptosDB {
     system_store: Arc<SystemStore>,
     transaction_store: Arc<TransactionStore>,
     pruner_config: StoragePrunerConfig,
-    pruner: Option<Pruner>,
+    state_pruner: Arc<Option<Box<dyn PrunerManager + Send + Sync>>>,
```
Comment (msmouse) [inline storage/aptosdb/src/lib.rs:267]: 1. `Option<StatePrunerManager>` should do?
2. Not introduced by this, but: imagine when we disable the pruner after being enabled for a while, the history is already incomplete, but with the current design we will assume we have complete history -- can we always initialize the pruners, but optionally enable the workers depending on config? This way we always know the min_readable_version. Also avoids a lot of `if Some(pruner) = ..`

File: storage/aptosdb/src/lib.rs [hunk 2]
```diff
@@ -167,23 +171,18 @@ fn error_if_too_many_requested(num_requested: u64, max_allowed: u64) -> Result<(
 }
 
 fn error_if_version_is_pruned(
-    pruner: &Option<Pruner>,
-    pruner_index: PrunerIndex,
+    pruner: &(dyn PrunerManager + Send + Sync),
```
Comment (msmouse) [inline storage/aptosdb/src/lib.rs:174]: Doesn't need to be `Send`? 

and nit: Shall we just make `trait PrunerManager: Sync` and avoid repeating it in places like this? like https://github.com/aptos-labs/aptos-core/blob/da86b2f29d67694713584e8b9b92f0b467af5f26/storage/state-view/src/lib.rs#L22

File: storage/aptosdb/src/lib.rs [hunk 3]
```diff
@@ -167,23 +171,18 @@ fn error_if_too_many_requested(num_requested: u64, max_allowed: u64) -> Result<(
...
-        {
-            ensure!(
-                version >= min_readable_version,
-                "{} version {} is pruned, min available version is {}.",
-                data_type,
-                version,
-                min_readable_version
-            );
-        }
+    if let Some(min_readable_version) = pruner.get_min_readable_version() {
```
Comment (msmouse) [inline storage/aptosdb/src/lib.rs:178]: I'm a bit confused what it means when min_readable_version is None... could you remind me about it? I feel it's ideal if we define that precisely.

Comment (zcchahaha) [reply storage/aptosdb/src/lib.rs:178]: I think it is none when the pruner is disabled.

File: storage/aptosdb/src/lib.rs (hunk 1 above)
Comment (zcchahaha) [reply storage/aptosdb/src/lib.rs:267]: 1. `error_if_version_is_pruned` will take either ledger pruner or state pruner so want to make it generic.
2. Can we do it in a separate PR after this refactoring PR?

Comment (msmouse) [reply storage/aptosdb/src/lib.rs:267]: 1. `fn fun(ref: &dyn Tr)` will work `fun(&ty)` if ty implements `Tr`

File: storage/aptosdb/src/lib.rs (hunk 3 above)
Comment (msmouse) [reply storage/aptosdb/src/lib.rs:178]: Not the case any more, right? You either don't have `pruner` to call the method on at all if you keep the `Option<XPrunerManager>` approach, or have the min_readable_version initialized in the `manager always there with optional worker` approach.

File: storage/aptosdb/src/lib.rs (hunk 1 above)
Comment (zcchahaha) [reply storage/aptosdb/src/lib.rs]: Addressed 1. Will have a follow up PR to do 2.

File: storage/aptosdb/src/lib.rs (hunk 3 above)
Comment (zcchahaha) [reply storage/aptosdb/src/lib.rs:178]: Will follow up PR to address this.

File: storage/aptosdb/src/lib.rs (hunk 2 above)
Comment (zcchahaha) [reply storage/aptosdb/src/lib.rs]: Done

File: storage/aptosdb/src/lib.rs [hunk 4]
```diff
@@ -993,9 +995,9 @@ impl DbReader for AptosDB {
     /// Get the first version that txn starts existent.
     fn get_first_txn_version(&self) -> Result<Option<Version>> {
         gauged_api("get_first_txn_version", || {
-            if let Some(pruner) = self.pruner.as_ref() {
+            if let Some(pruner) = self.state_pruner.as_ref() {
```
Comment (grao1991) [inline storage/aptosdb/src/lib.rs:998]: should this be ledger_pruner?

File: storage/aptosdb/src/lib.rs [hunk 5]
```diff
@@ -1005,9 +1007,9 @@ impl DbReader for AptosDB {
     /// Get the first version that write set starts existent.
     fn get_first_write_set_version(&self) -> Result<Option<Version>> {
         gauged_api("get_first_write_set_version", || {
-            if let Some(pruner) = self.pruner.as_ref() {
+            if let Some(pruner) = self.state_pruner.as_ref() {
```
Comment (grao1991) [inline storage/aptosdb/src/lib.rs:1010]: ditto

File: storage/aptosdb/src/lib.rs [hunk 6]
```diff
@@ -1391,8 +1378,8 @@ impl DbReader for AptosDB {
     fn get_ledger_prune_window(&self) -> Result<Option<usize>> {
         gauged_api("get_ledger_prune_window", || {
             let mut pruner_window = None;
-            if let Some(pruner) = self.pruner.as_ref() {
-                if let Some(window) = pruner.get_ledger_pruner_window() {
+            if let Some(pruner) = self.state_pruner.as_ref() {
```
Comment (grao1991) [inline storage/aptosdb/src/lib.rs:1381]: ditto

File: storage/aptosdb/src/lib.rs (hunk 4 above)
Comment (zcchahaha) [reply storage/aptosdb/src/lib.rs]: Thanks for the catch

**Comment by zcchahaha** [comment]
Will implemente`always initialize pruner but only spawn a worker thread if pruner enabled` approach in a separate PR.

File: storage/aptosdb/src/lib.rs [hunk 7]
```diff
@@ -1157,12 +1153,9 @@ impl DbReader for AptosDB {
         version: Version,
     ) -> Result<Option<StateValue>> {
         gauged_api("get_state_value_by_version", || {
-            error_if_version_is_pruned(
-                &self.pruner,
-                PrunerIndex::StateStorePrunerIndex,
-                "State",
-                version,
-            )?;
+            if let Some(state_pruner) = &self.state_pruner {
```
Comment (msmouse) [inline storage/aptosdb/src/lib.rs:1156]: point

**Review by msmouse — APPROVED** [review]

File: storage/aptosdb/src/lib.rs (hunk 7 above)
Comment (zcchahaha) [reply storage/aptosdb/src/lib.rs:1156]: Discussed offline. This is not introduced from the refactor and will be addressed in https://linear.app/aptoslabs/issue/BLK-82/%5Bdb%5D-state-kvs-prune-with-the-ledger-not-the-state-tree.

**Review by grao1991 — APPROVED** [review]

**Pull request merged**